datapackage_pipelines_ckan/processors/add_ckan_resource.py

```diff
@@ -0,0 +1,56 @@
+import os
+import json
+
+import requests
+
+from datapackage_pipelines.utilities.resources import (
+    PATH_PLACEHOLDER, PROP_STREAMED_FROM
+)
+from datapackage_pipelines.generators import slugify
+from datapackage_pipelines.wrapper import ingest, spew
+
+import logging
+log = logging.getLogger(__name__)
+
+parameters, datapackage, res_iter = ingest()
+
+ckan_host = parameters.pop('ckan-host')
+ckan_api_key = parameters.pop('ckan-api-key', None)
+request_headers = {}
+if ckan_api_key:
+    if ckan_api_key.startswith('env:'):
+        ckan_api_key = os.environ.get(ckan_api_key[4:])
+    request_headers.update({'Authorization': ckan_api_key})
+resource_id = parameters.pop('resource-id')
+resource_show_url = '{ckan_host}/api/3/action/resource_show'.format(
+                    ckan_host=ckan_host)
+
+response = requests.get(resource_show_url, params=dict(id=resource_id),
+                        headers=request_headers)
+
+try:
+    resource = response.json()['result']
+except json.decoder.JSONDecodeError:
+    log.error('Expected JSON in response from: {}'.format(resource_show_url))
+    raise
+
+if 'name' in resource:
+    if 'title' not in resource:
+        resource['title'] = resource['name']
+    resource['name'] = slugify(resource['name']).lower()
+
+if 'format' in resource:
+    resource['format'] = resource['format'].lower()
+
+if 'url' in resource:
+    resource['path'] = PATH_PLACEHOLDER
```

Comment:
Shouldn't this be set to `resource['url']`? Or does this mean that the actual file will be downloaded next to the resulting data package descriptor (`.`)?

Reply:
'path' is used as the output path for the resource when using dump.to_path. If we use the 'url', this creates a long path created from the url split at the '/' (http://www.newcastle.gov.uk/sites/drupalncc.newcastle.gov.uk/files/wwwfileroot/your-council/local_transparency/january_2012.csv). Instead we use `dpp:streamedFrom` to tell the pipeline where to fetch the resource data.

The next step in the example pipeline `stream_remote_resources` will do the actual streaming of remote resources, and `dump.to_path` will then create the following tree:

```
├── example
│   ├── data
│   │   └── january-2012.csv
│   └── datapackage.json
```

Reply:
That makes sense. So in the final `datapackage.json`, `resource['path']` will be `data/january-2012.csv`?

Reply:
That's right.

Reply:
Mid-pipeline, datapackages and resources can be a bit weird, and non-compliant with the spec, because they're not necessarily cooked yet :) By the time they hit a `dump.to_*` they should be though.

Reply:
OK, that is a bit confusing yes.